drivers/crypto/crypto_intel_sha.c

```diff
@@ -0,0 +1,334 @@
+	self->dfsha->pibcs.part.teie = 0;
+	/* set processing element enable */
+	self->dfsha->pibcs.part.peen = 1;
+
+	if (self->dfsha->shactl.part.en)
+		return -EINVAL; /* already enabled */
+
+	self->dfsha->shactl.part.hrsm = state_u.part.hrsm;
+
+	/* set initial values if resuming */
+	if (state_u.part.hrsm) {
+		err = intel_sha_set_resume_length_dw0(self, self->dfsha->shaaldw0.full);
+		if (err)
+			return err;
+		err = intel_sha_set_resume_length_dw1(self, self->dfsha->shaaldw1.full);
+		if (err)
+			return err;
+		err = intel_sha_regs_cpy((void *)self->dfsha->initial_vector,
+					 (void *)self->dfsha->sha_result,
+					 sizeof(self->dfsha->initial_vector));
+		if (err)
+			return err;
+	}
+
+	/* set ctl hash first middle */
+	if (self->dfsha->shactl.part.en)
+		return -EINVAL; /* already enabled */
+
+	self->dfsha->shactl.part.hfm = state_u.part.state;
+
+	/* increment pointer */
+	self->dfsha->pibfpi.full = buf_in_size;
+
+	err = intel_sha_set_ctl_enable(self, 1);
+	if (err)
+		return err;
+
+	err = intel_sha_set_ctl_enable(self, 0);
+
+	return err;
+}
+
+static int intel_sha_copy_hash(struct sha_container *const self, void *dst, size_t len)
+{
+	/* NOTE: generated hash value should be read from the end */
+
+	int err = -EINVAL;
+	uint32_t counter = 0;
+	uint32_t last_idx = 0;
+
+	if ((IS_ALIGNED(len, sizeof(uint32_t))) && (IS_ALIGNED(dst, sizeof(uint32_t)))) {
+		len /= sizeof(uint32_t);
+		counter = 0;
+		/* The index of a last element in the sha result buffer. */
+		last_idx = (sizeof(self->dfsha->sha_result) / sizeof(uint32_t)) - 1;
+
+		for (counter = 0; counter != len; counter++) {
+			((uint32_t *)dst)[counter] =
+				((uint32_t *)self->dfsha->sha_result)[last_idx - counter];
+		}
+
+		err = 0;
+	}
+
+	return err;
+}
+
+static int intel_sha_device_get_hash(struct device *dev, void *buf_out, size_t buf_out_size)
+{
+	int err;
+	struct sha_container *const self = dev->data;
+
+	if (dev == NULL)
+		return -ENODEV;
+	if (buf_out == NULL)
+		return -EINVAL;
+	/* wait until not busy */
+	while (self->dfsha->shasts.part.busy) {
+	}
+
+	err = intel_sha_copy_hash(self, buf_out, buf_out_size);
+
+	return err;
+}
+
+static int intel_sha_compute(struct hash_ctx *ctx, struct hash_pkt *pkt, bool finish)
+{
+	xthal_dcache_region_writeback((uint8_t *)(pkt->in_buf), pkt->in_len);
+
+	int ret;
+	struct sha_container *self = (struct sha_container *const)(ctx->device)->data;
+	struct sha_session *session = (struct sha_session *)ctx->drv_sessn_state;
+	size_t frag_length;
+	size_t output_size;
+	uint32_t *hash_int_ptr = (uint32_t *)(pkt->out_buf);
+
+	/* set algo */
+	self->dfsha->shactl.full = 0x0;
+	self->dfsha->shactl.part.algo = session->algo;
+
+	/* restore ctx */
+	self->dfsha->shaaldw0 = session->sha_ctx.shaaldw0;
+	self->dfsha->shaaldw1 = session->sha_ctx.shaaldw1;
+
+	ret = intel_sha_regs_cpy((void *)self->dfsha->initial_vector,
+				 (void *)session->sha_ctx.initial_vector,
+				 sizeof(self->dfsha->initial_vector));
+	if (ret)
+		return ret;
```

Comment:
```suggestion
	if (ret)
		
```
Zephyr's code style requires braces for this kind of statement.